compiler/src/iree/compiler/DispatchCreation/SplitReduction.cpp

```diff
@@ -24,14 +25,355 @@ namespace mlir::iree_compiler::DispatchCreation {
+                   "with a reduction dimension of 128."),
+    llvm::cl::init(true));
+
+// Controls the tile size used when applying split-k to argmax reductions.
+// This value defines how many elements along the reduction dimension are
+// processed per tile (e.g., a value of 128 means each tile reduces over 128
+// elements).
+static llvm::cl::opt<int64_t> splitArgmaxReductionRatio(
+    "iree-dispatch-creation-split-argmax-reduction",
+    llvm::cl::desc("Ratio to split argmax. Set to 0 or 1 to disable"),
+    llvm::cl::init(1));
 
 static llvm::cl::list<int64_t> topkSplitReductionRatio(
     "iree-dispatch-creation-topk-split-reduction",
     llvm::cl::desc("comma separated list of split ratios"),
     llvm::cl::CommaSeparated);
 
+template <typename LinalgOpTy>
+static FailureOr<linalg::SplitReductionResult>
+splitReductionImpl(RewriterBase &rewriter, LinalgOpTy op,
+                   linalg::ControlSplitReductionFn controlSplitReductionFn) {
+  return linalg::splitReduction(rewriter, op, controlSplitReductionFn);
+}
+
+// Matches the combiner pattern in a linalg.generic argmax-style reduction:
+// Example MLIR:
+// %4:2 = linalg.generic {
+//     indexing_maps = [...],
+//     iterator_types = ["parallel", "reduction"]
+//   } ins(%arg0 : tensor<?x128xbf16>) outs(%1, %3 : tensor<?xbf16>,
+//   tensor<?xi64>) {
+// ^bb0(%in: bf16, %out: bf16, %out_0: i64):
+//   %5 = linalg.index 1 : index
+//   %6 = arith.index_cast %5 : index to i64
+//   %7 = arith.maximumf %in, %out : bf16
+//   %8 = arith.cmpf ogt, %in, %out : bf16
+//   %9 = arith.select %8, %6, %out_0 : i64
+//   linalg.yield %7, %9 : bf16, i64
+// } -> (tensor<?xbf16>, tensor<?xi64>)
+//
+// This function extracts the `arith.maximumf`, `arith.cmpf`, and `arith.select`
+// operations from the body to facilitate transformations such as split
+// reduction.
+static LogicalResult
+collectArgmaxCombinerOps(linalg::GenericOp genericOp,
+                         MutableArrayRef<Operation *> combinerOps) {
+  if (combinerOps.size() < 3) {
+    return genericOp->emitError(
+        "combinerOps must have space for exactly 3 elements");
+  }
+
+  assert(IREE::LinalgExt::isArgmaxOp(genericOp) &&
+         "expected operation to be an argmax op");
+
+  auto yieldOp = cast<linalg::YieldOp>(genericOp.getBody()->getTerminator());
+
+  // Extract max value producer: arith.maximumf.
+  Value maxResult = yieldOp.getOperand(0);
+  auto maxOp = dyn_cast<arith::MaximumFOp>(maxResult.getDefiningOp());
+
+  // Extract index result producer: arith.select.
+  Value indexResult = yieldOp.getOperand(1);
+  auto selectOp = dyn_cast<arith::SelectOp>(indexResult.getDefiningOp());
+
+  // Extract the condition of the select, expected to be arith.cmpf with
+  // predicate OGT.
+  auto cmpOp = dyn_cast<arith::CmpFOp>(selectOp.getCondition().getDefiningOp());
+
+  combinerOps[0] = maxOp;
+  combinerOps[1] = selectOp;
+  combinerOps[2] = cmpOp;
```

Comment:
I would rather create a struct combinerOps with Operation* max, ..., so you can directly query the maxOp and don't care about the extra bookkeeping.